_source/_posts/2022-08-15-spring-boot-crud-update.md

```diff
@@ -0,0 +1,1077 @@
+
+{% img blog/spring-boot-crud-update/spring-and-vue.png alt:"Spring Boot, Vue, and Okta logos" width:"500" %}{: .center-image }
+
+This project has two major parts:
+
+- Spring Boot server
+- Vue client
+
+The Spring Boot server will include an H2 in-memory database and will use Spring Data JPA to map our todo data model to a database table for persistence. As you'll see, the server will leverage Spring Boot's ability to quickly expose data via a REST API with a minimum of configuration.
+
+The client will use [Vue 3](https://vuejs.org/) and the Quasar Framework. [The Quasar Framework](https://quasar.dev/) provides components and layout tools to help build Vue applications quickly with a consistent, high-quality user interface.
+
+**Prerequisites:**
+
+- [Java 11](https://adoptium.net/): or use [SDKMAN!](https://sdkman.io/) to manage and install multiple version
+
+- [Okta CLI](https://cli.okta.com/manual/#installation)
+
+- [HTTPie](https://httpie.org/doc#installation)
+
+- [Node 16+](https://nodejs.org)
+- [Vue CLI](https://cli.vuejs.org/guide/installation.html)
+
+You will need a free Okta Developer account if you don't already have one. But you can wait until later in the tutorial and use the Okta CLI to login or register for a new account.
+
+## Create Okta OIDC app
+
+Open a Bash shell. Create a parent directory for the project. Eventually, this will include both the resource server and client projects.
+
+```bash
+mkdir spring-boot-vue-crud
+cd spring-boot-vue-crud
+```
+
+If you already have an Okta account, log in using `okta login`. If you need to register for a new, free account: `okta register`.
+
+Create an OIDC app using the following command:
+
+```bash
+okta apps create
+```
+
+**Application name**: `spring-boot-vue-crud`
+
+**Type of Application**: `2: Single Page App`
+
+**Redirect URI(s)**: `http://localhost:8080/login/callback`
+
+All other defaults are fine.
+
+Your console should finish with something like this:
+
+```bash
+...
+Configuring a new OIDC Application, almost done:
+Created OIDC application, client-id: 0oa5rdbh01J0293u209
+|
+Okta application configuration: 
+Issuer:    https://dev-123567.okta.com/oauth2/default
+Client ID: 0oa5rdbh01J0293u209
+```
+
+Copy the client ID and issuer URI somewhere. You'll need them for both the client and resource server applications.
+
+## Bootstrap the Spring Boot app using Spring Initializr
+
+You're going to use [the Spring Initializr](start.spring.io/) to create a starter project for the resource server. You can take a look at the project website if you want, but here you're going to use the REST API to download a pre-configured starter.
+
+The following command will download the starter project and un-tar it to a new directory named `resource-server`.
+
+```bash
+curl https://start.spring.io/starter.tgz \
+  -d bootVersion=2.6.10 \
```

Comment:
Use 2.7.2, a 3.0 milestone, or no version at all. Spring Boot seems to be better than most at maintaining backward compatibility between major versions.

Reply:
I tried that. I get this error:
```bash
curl https://start.spring.io/starter.tgz \
  -d bootVersion=2.7.2 \
  -d jvmVersion=11 \
  -d dependencies=web,data-rest,lombok,data-jpa,h2,okta \
  -d type=gradle-project \
  -d baseDir=resource-server
{"timestamp":"2022-08-17T18:24:25.088+00:00","status":400,"error":"Bad Request","message":"Dependency 'okta' is not compatible with Spring Boot 2.7.2","path":"/starter.tgz"}
```

Reply:
@bdemers Do you know how to get this fixed on start.spring.io?

Reply:
I created an issue to track and fix: https://github.com/okta/okta-spring-boot/issues/462

Reply:
@moksamedia I think it'd be good to use a two-step approach in this post: 

1. Clone a repo, configure the app, run it, then explain it.
2. Show how to create from scratch.

This makes it easier to maintain because the "create from scratch" steps will stop working after a couple of years. This is because start.spring.io removes versions.

Also, I just realized my recent update to [Spring Boot React](https://developer.okta.com/blog/2022/06/17/simple-crud-react-and-spring-boot) doesn't use the clone a repo approach, so you can do whatever you like. ;-)